Execution control for Table.write and various widget tweaks...

jdunkerley opened the pull request:
### Pull Request Description

- Adds execution control to `Table.write`.
  - Refactored the `Text.write` to make part reusable.
  - Tidied up some legacy mess in tests.
- Add easier flow to go from `Text` to an `URI` to fetching data.
- Add decode functions to `Response` and `Response_Body`.
- Fix issue with 0 length regex matches (using same as Python and .Net approach).
- Add various ALIAS entries to make function discovery easier.
- Sort a lot of drop down and vector editors out (including switch to fully qualified names).

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [x] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [ ] If GUI codebase was changed, the GUI was tested when built using `./run ide build`.

jdunkerley committed:
Restructure `Text.write` and reuse components to control `Table.write`.

jdunkerley committed:
Add dropdown for `aggregate`.
Remove dead argument from `select_columns`, `remove_columns` and `reorder_columns`.
Add ALIAS to `remove_columns`.

jdunkerley committed:
Add ALIAS for reading sheets and ranges.
Add custom drop down for `keep_unmatched`.

jdunkerley committed:
Fix issue where `tables` didn't allow a single value.

jdunkerley committed:
Fix not returning the File.

jdunkerley committed:
Fix not returning the File.

jdunkerley committed:
Simplify docs a bit.

jdunkerley committed:
Fix for find_all.
Add fetch to URI.
Add some ALIASes (+, -, few methods).

jdunkerley committed:
Add some ALIASes (+, -, few methods).

jdunkerley committed:
ALIASes for `*` and `/`.

jdunkerley committed:
Add all the other ALIASes for operators.

jdunkerley committed:
Remove legacy default_widget for Filter_Condition.
Use fully qualified name for Encoding.

jdunkerley committed:
Use fully qualified name for Locale dropdowns.
Use fully qualified names for File_Format dropdowns.
Use fully qualified names for Report_Unmatched dropdowns.
Use fully qualified names for aggregates and parse type dropdowns.

jdunkerley committed:
Adding decode to Response and Response_Body.

radeusgd commented on distribution/lib/Standard/Table/0.0.0-dev/src/Internal/Widget_Helpers.enso:
```diff
@@ -16,40 +18,48 @@ import project.Data.Table_Conversions
+    count_not_nothing = Option "Count Not Nothing" fqn+".Count_Not_Nothing" [column_widget]
+    count_nothing = Option "Count Nothing" fqn+".Count_Nothing" [column_widget]
 
     ## Should be a list of Text columns only
```
Should we start implementing this by-type column filtering? We have all the capabilities needed to do that.

If not in this PR, I'd be happy to create a small PR for this - it's just a few lines of adjustment to get a really nice UX improvement.

jdunkerley replied:
Discussed with Ned and he argued against it.

radeusgd replied:
Oh, why?

jdunkerley replied:
Logic was more confusing for columns to vanish depending on the mode.

radeusgd replied:
Hmm, I see. Well indeed maybe it would be less confusing if we chose the column _first_ and then had a list of aggregations available for it - but that does not fit our `Aggregate_Column` format.

IMHO suggesting invalid columns for an aggregation is not that helpful, I don't imagine it is a problem, but if that's the recommendation.

jdunkerley replied:
I was thinking the same and wondered around adding a note if an unsupported type - so you would see the same set but something in brackets warning invalid for this).

radeusgd replied:
Yes, I think that would be good. Or even, if we had the ability to add styling, the invalid types could be greyed out.

Also, I would move all unsupported down to the bottom, so that the first fields that show up are the supported ones. No sense in having to scroll thru unsupported fields to find a good one.

radeusgd replied:
Shall we create a follow up task for this?

I think the sorting will be especially useful, and marking the not-matching ones should also be helpful. Since we don't have formatting yet, maybe we can mark them like `/Shortest/`?

jdunkerley replied:
Will add a task.

jdunkerley committed:
Align find_all with C#/python approach.

jdunkerley committed:
Remove context imports which aren't used.
New test for disabled writing of tables.

radeusgd commented on distribution/lib/Standard/Base/0.0.0-dev/src/Data/Boolean.enso:
```diff
@@ -15,7 +15,8 @@ type Boolean
 
-    ## Computes the logical and (conjunction) of two booleans.
+    ## ALIAS And
+       Computes the logical and (conjunction) of two booleans.
```
```suggestion
       Computes the conjunction (logical 'and') of two booleans.
```

I'm wondering if that form would not be more readable? (low priority)

radeusgd commented on distribution/lib/Standard/Base/0.0.0-dev/src/Data/Locale.enso:
```diff
@@ -445,5 +445,6 @@ type Locale
     widget_options = Locale.predefined_locale_fields.map field_name->
         display_string = field_name.replace '_' ' ' . to_case (if field_name.length == 2 then Case.Upper else Case.Title)
-        code_string = "Locale." + field_name
+        fqn = Meta.get_qualified_type_name Locale
+        code_string = fqn + "." + field_name
```
Why do we need a FQN now?

jdunkerley replied:
Requested by IDE team - the widgets should auto resolve now.

radeusgd commented on distribution/lib/Standard/Base/0.0.0-dev/src/Data/Numbers.enso:
```diff
@@ -433,7 +431,8 @@ type Decimal
     / self that = @Builtin_Method "Decimal./"
 
-    ## Computes the remainder when dividing this by that.
+    ## ALIAS Modulus
```
```suggestion
    ## ALIAS Modulus, Modulo
```

Maybe also this form? It's pretty common, especially coming from Polish, but I see it used in English too.

radeusgd commented on distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Regex/Pattern.enso:
```diff
@@ -352,8 +353,8 @@ type Match_Iterator
     next self =
-        regex_result = self.pattern.internal_regex_object.exec self.input self.cursor
-        case regex_result.isMatch of
+        regex_result = if self.cursor > self.input.char_vector.length then Nothing else self.pattern.internal_regex_object.exec self.input self.cursor
```
Why it's not?
```suggestion
        regex_result = if self.cursor >= self.input.char_vector.length then Nothing else self.pattern.internal_regex_object.exec self.input self.cursor
```

I assume you already tested it well enough, there is a good reason, so I'm not suggesting to change. Just would like to understand it.

If it's `== length` it's already after last char so seems enough to end?

jdunkerley replied:
`== length` gets you the empty match to meet the bits.
`"".find_all ".*"` starts at cursor 0 and returns a match of `""`

radeusgd commented on distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Regex/Pattern.enso:
```diff
@@ -363,7 +364,8 @@ type Match_Iterator
                 match = Match.Value self.pattern regex_result self.input
-                next_cursor = match.utf_16_end 0
+                ## Handle edge case where match is 0 length
+                next_cursor = (self.cursor + 1).max (match.utf_16_end 0)
```
nitpick: I prefer this syntax, IMO its more readable.

```suggestion
                next_cursor = Math.max (self.cursor + 1) (match.utf_16_end 0)
```

Also the Enso comments actually encourage it :)

```
   ? Math.max or Number.max
     While we provide the max method on `Number`, we find it more intuitive to
     write `Math.max a b` rather than `a.max b`. To that end, we recommend using
     the first style.
```

jdunkerley committed:
CHANGELOG.

radeusgd commented on distribution/lib/Standard/Base/0.0.0-dev/src/Network/HTTP/Response.enso:
```diff
@@ -30,6 +40,13 @@ type Response
+    content_type self =
+        response_headers = self.headers
+        content_type = response_headers.find if_missing=Nothing h-> "Content-Type".equals_ignore_case h.name
+        if content_type.is_nothing then Nothing else content_type.value
```
not for now, but I'm wondering if we should look for some idiom for this, like Kotlin's elvis operator:
```suggestion
        content_type?.value
```

jdunkerley replied:
yeah would be a nice addition I think.

radeusgd commented on distribution/lib/Standard/Base/0.0.0-dev/src/Network/HTTP/Response.enso:
```diff
@@ -30,6 +40,13 @@ type Response
+    content_type : Text | Nothing
+    content_type self =
+        response_headers = self.headers
+        content_type = response_headers.find if_missing=Nothing h-> "Content-Type".equals_ignore_case h.name
```
Maybe that name would be clearer?

```suggestion
        content_type_header = response_headers.find if_missing=Nothing h-> "Content-Type".equals_ignore_case h.name
```

radeusgd replied:
Also, that is a pretty tedious way to figure out the content type.

I guess its simple code and probably not a preformance intensive place, so we can leave as-is.

But instead of constructing the whole Header Map in Enso and copying all headers around, we could just use the Java object directly:
```suggestion
        java_optional = self.internal_http_response.headers.firstValue "Content-Type"
        if java_optional.isPresent then java_optional.get else Nothing
```

jdunkerley committed:
Fix Formats test.

radeusgd commented on distribution/lib/Standard/Base/0.0.0-dev/src/Network/HTTP/Response.enso:
```diff
@@ -54,6 +71,41 @@ type Response
+    @format File_Format.default_widget
+    decode : File_Format -> Any -> Any
+    decode self format=Auto_Detect ~if_unsupported=(Error.throw (Unsupported_Content_Type.Value self.content_type)) =
+        if format != Auto_Detect then format.read_web self else
```
Not all formats define a `read_web` method. They tend to define `for_web`, but `read_web` is only on some.

If the format is specified by the user and it does not define `read_web`, it will fail with a missing method exception that is not friendly to the user.

Shall we add the missing `read_web` to all formats? IMO it should be present everywhere and just in some cases all it does may be throw "this is unsupported". That way the errors will be more friendly.

Alternatively we can use Meta/catch the Panic and verify if this is supported at the caller level.

Anyway, I think we should try to have better errors than `No_Such_Method`. Fine to do in a follow up PR, but let's not forget about it.

radeusgd commented on distribution/lib/Standard/Base/0.0.0-dev/src/Network/HTTP/Response.enso:
```diff
@@ -69,3 +121,19 @@ type Response
+
+       Arguments:
+       - content_type: The content type that is unsupported.
+    Value content_type
```
What is the type here? From the `to_display_text` I conclude it is `Text | Nothing`. Can we specify it in the constructor? Makes the code easier to read.

radeusgd replied:
```suggestion
    Value (content_type : Text | Nothing)
```

radeusgd replied:
Shouldn't this be called error or problem?

```suggestion
    Error content_type
```

I see it is an error that is thrown around. The `Value` constructor usually is not used for those.

radeusgd replied:
```suggestion
    Error (content_type : Text | Nothing)
```

radeusgd commented on distribution/lib/Standard/Base/0.0.0-dev/src/Network/HTTP/Response_Body.enso:
```diff
@@ -18,10 +19,31 @@ type Response_Body
+        self.decode_as_text encoding . parse_json
+
     ## PRIVATE
        Convert response body to Text.
     to_text : Text
-    to_text self = Text.from_utf_8 self.bytes
+    to_text self = self.decode_as_text Encoding.utf_8
 
```
Tbh I don't like that `to_text` may fail.

This is actually problematic, because `to_text` is the delegate called when this object is passed to Java polyglot and `toString` is called on it.

When it returns a value with warnings or a dataflow error, I'm not sure what happens but I expect weird stuff may.

IMO we should just always use `decode_` to do decoding and this should be a short representation like `Response_Body ...`. Alternatively, may be worth double checking what happens if `to_text` returns warnings/dataflow-error and `toString` is called thru polyglot - I may have a test tomorrow. Maybe we should check for errors/warnings and then replace with some message "Response_Body that cannot be decoded in UTF-8, use decode_as_text with other encoding to get the contents"?

jdunkerley committed:
Add Modulo alias.
Use `Math.max`.

radeusgd commented on distribution/lib/Standard/Table/0.0.0-dev/src/Data/Table.enso:
```diff
@@ -1953,8 +1956,14 @@ type Table
+                if methods.contains "write_table" . not then Error.throw (File_Error.Unsupported_Output_Type format Table) else
+                    effective_existing_behaviour = on_existing_file.get_effective_behavior file
+                    tgt_file = if Context.Output.is_enabled then file else
+                        file.create_dry_run_file copy_original=on_existing_file==Existing_File_Behavior.Append
```
I'm just not so sure of operator precedence. I assume its correct but it spawns fear in my eyes.
Maybe thiswould be more readable?

```suggestion
                        should_copy = on_existing_file==Existing_File_Behavior.Append
                        file.create_dry_run_file copy_original=should_copy
```

radeusgd commented on distribution/lib/Standard/Table/0.0.0-dev/src/Excel/Excel_Workbook.enso:
```diff
@@ -103,10 +103,15 @@ type Excel_Workbook
     @types (self-> Single_Choice values=(self.table_types.map t-> Option t t.pretty))
     tables : Text -> Text -> Text -> Vector -> Boolean -> Table
     tables self name_like=Nothing database=self.database schema=self.schema types=Nothing all_fields=False =
+        types_array = case types of
```
```suggestion
        types_vector = case types of
```

This name is confusing. Array is a different type from Vector after all

radeusgd commented on test/Table_Tests/src/IO/Formats_Spec.enso:
```diff
@@ -79,4 +79,40 @@ spec = Test.group 'Various File Format support on Table' <|
+            s.exists.should_be_true
+
+            r_data = s.read
+            row_count = if r_data . is_a Table then r_data.row_count else r_data.length
```
When is `r_data` not a `Table` here??

jdunkerley replied:
Writing to a JSON file gets read back in as a Vector by default.

radeusgd reviewed (approved):
Looks good to me.

I have some comments/suggestions. Nothing blocking, if we want to get this merged, I think it's fine to leave them for a follow-up PR. Just please let's not forget them :)

jdunkerley committed:
Use Java Optional for Content-Type.
Make Response_Body to_text more stable.
Add local variable for copying to temp file.

jdunkerley committed:
Rename to types_vector.

GregoryTravis commented on distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Regex/Pattern.enso:
```diff
@@ -97,7 +99,7 @@ type Pattern
 
        Arguments:
        - input: The text to match the pattern described by `self` against.
```
This could use a brief comment explaining the special handling of zero-length matches.

jdunkerley committed:
Fix issues with decode.

jdunkerley committed:
Fix issues with format types and fetch methods.

jdunkerley committed:
Better `on` for `join`.

jdunkerley committed:
Fix failed test.
